onnxruntime/core/graph/graph_utils.h

```diff
@@ -86,20 +81,99 @@ bool GetRepeatedNodeAttributeValues(const Node& node,
+
+After output edges are updated, remove the node.
+*/
 bool RemoveNode(Graph& graph, Node& node);
 
+/** Tests if we can remove a node and replace its output with an initializer.
+Conditions:
+ - Only one of the node's outputs is used by downstream operators or as a graph output 
+   - multiple edges for the single used output are allowed
+ - If the node produces a graph output the initializer_name must be the same as the node's output name
+   - otherwise the required graph output will not be produced
+ - Removing the node won't break a subgraph that consumes the node's output
+*/
+bool CanReplaceNodeWithInitializer(const Graph& graph, const Node& node, const std::string& initializer_name);
+
+/** Remove a node and replace its output with the provided NodeArg for an initializer.
+See CanReplaceNodeWithInitializer for the conditions that must be satisfied in order to remove the node.*/
+bool ReplaceNodeWithInitializer(Graph& graph, Node& node, NodeArg& replacement);
+
 /** Removes all output edges from the given Node of the Graph. 
     This should probably be elevated to the Graph API eventually. */
 size_t RemoveNodeOutputEdges(Graph& graph, Node& node);
 
+/** Replaces a single output of a node, with an output from a different node.
+Moves the output edges from 'node' for 'output_idx' to the replacement node.
+@param replacement The node providing the replacement output.
+@param replacement_output_idx The index of the output from 'replacement' to use. 
+
+e.g. Node A produces outputs A1 and A2. 
+     Node B consumes A2 (edge between A and B for A2) and produces B1. 
+     Node C consumes B1 (edge between B and C for B1).
+
+     If Node B was determined to not be needed, you would call ReplaceNodeOutput(graph, B, 0, A, 1) 
+     to replace B1 (output index 0 for node B) with A2 (output index 1 for node A).
+     The edge that existed between B and C for B1 with be removed, and replaced with an edge between A and C for A2.
+*/
+void ReplaceNodeOutput(Graph& graph, Node& node, int output_idx, Node& replacement, int replacement_output_idx);
+
+/** Replace the input to a node with a NodeArg for an initializer or graph input.
+@remarks There is no edge between an initializer or graph input and a Node, so the replacement only updates the 
+         node's input definition.
+*/
+void ReplaceNodeInput(Node& target, int target_input_idx, NodeArg& replacement);
+
+/** Add an input to a node with a NodeArg for an initializer or graph input.
+@remarks target_input_idx must be the next input slot. 
+           e.g. if a Node has 2 inputs, AddNodeInput can only add input 3 and not 4. 
+         There is no edge between an initializer or graph input and a Node, so the replacement only updates the 
+         node's input definition and does not create any new edges.
+*/
+void AddNodeInput(Node& target, int target_input_idx, NodeArg& replacement);
```

Comment:
replacement -> new_input?
(this is what you use in the implementation too)

Reply:
Updated